infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/TopologyDescriptor.java

```diff
@@ -48,81 +65,81 @@ public class TopologyDescriptor {
-     *
-     * @param topologyMessage   topology message
-     * @param localNodeId       the identifier of this node
-     */
-    public TopologyDescriptor(TopologyConfigurationMsg topologyMessage, String localNodeId) {
-        this.topologyConfigId = topologyMessage.getTopologyConfigID();
-        this.sinkClusters = new HashMap<>();
-        this.sourceClusters = new HashMap<>();
-        this.invalidClusters = new HashMap<>();
-
-        for (ClusterConfigurationMsg clusterConfig : topologyMessage.getClustersList()) {
-            ClusterDescriptor cluster = new ClusterDescriptor(clusterConfig);
-            if (clusterConfig.getRole() == ClusterRole.SOURCE) {
-                sourceClusters.put(cluster.getClusterId(), cluster);
-            } else if (clusterConfig.getRole() == ClusterRole.SINK) {
-                sinkClusters.put(cluster.getClusterId(), cluster);
-            } else {
-                invalidClusters.put(cluster.getClusterId(), cluster);
-            }
-        }
-
-        setLocalDescriptor(localNodeId);
-    }
 
     /**
      * Constructor
      *
-     * @param topologyConfigId topology configuration identifier (epoch)
-     * @param sourceClusters source cluster's
-     * @param sinkClusters sink cluster's
+     * @param topologyConfigId  topology ID
+     * @param localNodeId      the identifier of this node
+     * @param remoteSinkToReplicationModel  remote Sink to local cluster and corresponding replication models
+     * @param remoteSourceToReplicationModel  remote Source to local cluster and corresponding replication models
+     * @param remoteClusterEndpoints remote clusters to which the local cluster will initiate connection
+     * @param otherClusters      remote clusters which are neither Source nor Sink to the local cluster
      */
-    public TopologyDescriptor(long topologyConfigId, @NonNull List<ClusterDescriptor> sourceClusters,
-                              @NonNull List<ClusterDescriptor> sinkClusters, String localNodeId) {
+    public TopologyDescriptor(long topologyConfigId, String localNodeId,
+                              Map<ClusterDescriptor, Set<LogReplication.ReplicationModel>> remoteSinkToReplicationModel,
+                              Map<ClusterDescriptor, Set<LogReplication.ReplicationModel>> remoteSourceToReplicationModel,
+                              Set<ClusterDescriptor> remoteClusterEndpoints, Set<ClusterDescriptor> otherClusters) {
+
         this.topologyConfigId = topologyConfigId;
-        this.sourceClusters = new HashMap<>();
-        this.sinkClusters = new HashMap<>();
-        this.invalidClusters = new HashMap<>();
+        this.remoteSinkClusterToReplicationModels = remoteSinkToReplicationModel;
+        this.remoteSourceClusterToReplicationModels = remoteSourceToReplicationModel;
+        remoteClusterEndpoints.forEach(cluster -> this.remoteClusterEndpoints.put(cluster.getClusterId(), cluster));
+
+        this.remoteSinkClusters = new HashMap<>();
```

Comment:
shouldnt we isolate the local cluster so that it does not get added to these maps?

Reply:
No. the local cluster will not come in `remoteSinkToReplicationModel` or `remoteSourceToReplicationModel`.
It will be present in the `otherClusters`.
So no need to remove it from remote* structures.